java/WordCountTrove.java

```diff
@@ -0,0 +1,167 @@
+            retval.append(new String(word, StandardCharsets.UTF_8))
+                .append('\t')
+                .append(count);
+            return retval;
+        }
+    }
+
+    private static class ByteArrayComparator implements Comparator<byte[]> {
+        @Override
+        public int compare(byte[] b1, byte[] b2) {
+            // Sorting by byte value is less correct
+            // but avoids very expensive UTF-8 decoding
+            int limit = Math.min(b1.length, b2.length);
+            for(int i=0; i<limit; i++) {
+                int diff = b1[i] - b2[i];
+                if (diff != 0) {
+                    return diff;
+                }
+            }
+            return b1.length-b2.length;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return o == this;
+        }
+    }
+
+    private static final ByteArrayComparator BYTE_COMPARATOR_INSTANCE = new ByteArrayComparator();
+    
+    private static void submitWord(TObjectIntCustomHashMap<byte[]> m, byte[] word) {
+        m.adjustOrPutValue(word, 1, 1);        
+    }
+    
+    public static void main(String[] args) throws IOException {
+        System.err.println("Parsing and adding to map");
+        long startTime = System.currentTimeMillis();
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in, "UTF-8"));
+        TObjectIntCustomHashMap<byte[]> m = new TObjectIntCustomHashMap<byte[]>(new BytesHashingStrategy(),1000000, 0.75f, -1);
+        String line;
+        while ((line = br.readLine()) != null) {
+            line = line.trim();
+            if (!line.isEmpty()) {
+                int index = 0;
+                for(int i = 0; i < line.length(); i++){
+                    char c = line.charAt(i);
+                    if(c == '\t' || c == ' '){
+                        if (index == i) {
+                            index ++;
+                        } else {
+                            byte[] word = line.substring(index, i).getBytes(StandardCharsets.UTF_8);
+                            index = i + 1;
+                            submitWord(m, word);
+                        }
+                    }
+                }
+                if(index < line.length()){
+                    submitWord(m, line.substring(index).getBytes(StandardCharsets.UTF_8));
+                }
+            }
+        }
+        br.close();
+        long endTime = System.currentTimeMillis();
+        System.err.println("Parsing/map addition time (ms): "+(endTime-startTime));
+
+        System.err.println("Creating Count objects for sorting");
+        startTime = System.currentTimeMillis();
+
+        final ArrayList<CountForWord> multiples = new ArrayList<CountForWord>(m.size()/2);
+        final ArrayList<byte[]> singles = new ArrayList<byte[]>(m.size()/2);
+
+        TObjectIntProcedure<byte[]> proc = new TObjectIntProcedure<byte[]>(){
+            int i=0; 
```

Comment:
Heh, leftover junk from a previous implementation, oops!  (I was doing an array for performance reasons before, and needed an indexing variable).